File: python-example/non_bdd_tests.py
import unittest
from unittest.mock import patch, MagicMock
from username_validator import UsernameValidator

class NonBDDTests(unittest.TestCase):
    def test_is_valid(self):
        # Create a mock/spy for the validator
        validator = UsernameValidator()
        
        # Mock the methods to track calls
        validator.is_too_short = MagicMock(wraps=validator.is_too_short)
        validator.is_too_long = MagicMock(wraps=validator.is_too_long)
        validator.contains_illegal_chars = MagicMock(wraps=validator.contains_illegal_chars)
        
        username = "User@123"
        result = validator.is_valid(username)
        
        # Check if all methods were called with the right input
        validator.is_too_short.assert_called_once_with(username)
        validator.is_too_long.assert_called_once_with(username)
        validator.contains_illegal_chars.assert_called_once_with(username)
        
        # Now check if they return the correct results
        self.assertFalse(validator.is_too_short(username))
        self.assertFalse(validator.is_too_long(username))
        self.assertTrue(validator.contains_illegal_chars(username))

if __name__ == "__main__":
    unittest.main()

File: ruby-example/non_bdd_tests.rb
require 'minitest/autorun'
require 'mocha/minitest'
require_relative 'username_validator'

class NonBDDTests < Minitest::Test
  def test_is_valid
    # Create a mock/spy for the validator
    validator = UsernameValidator.new
    
    # Mock the methods to track calls
    validator.expects(:is_too_short).once.returns(false).with('User@123')
    validator.expects(:is_too_long).once.returns(false).with('User@123')
    validator.expects(:contains_illegal_chars).once.returns(true).with('User@123')
    
    username = 'User@123'
    result = validator.is_valid(username)
    
    # The mocked methods will automatically verify that they were called correctly
    # Additionally, we can assert the expected return values
    validator = UsernameValidator.new  # Create a fresh instance for actual return checking
    refute validator.is_too_short(username)
    refute validator.is_too_long(username)
    assert validator.contains_illegal_chars(username)
  end
end